frontend/ui/widget/imageviewer.lua

```diff
@@ -180,180 +167,348 @@ function ImageViewer:update()
-                        UIManager:close(self)
-                    end,
-                },
+                id = "scale",
+                text = self._scale_to_fit and _("Original size") or _("Scale"),
+                callback = function()
+                    self.scale_factor = self._scale_to_fit and 1 or 0
+                    self._scale_to_fit = not self._scale_to_fit
+                    -- Reset center ratio (may have been modified if some panning was done)
+                    self._center_x_ratio = 0.5
+                    self._center_y_ratio = 0.5
+                    self:update()
+                end,
             },
-        }
-        local button_table = ButtonTable:new{
-            width = self.width - 2*self.button_padding,
-            button_font_face = "cfont",
-            button_font_size = 20,
-            buttons = buttons,
-            zero_sep = true,
-            show_parent = self,
-        }
-        button_container = CenterContainer:new{
-            dimen = Geom:new{
-                w = self.width,
-                h = button_table:getSize().h,
+            {
+                id = "rotate",
+                text = self.rotated and _("No rotation") or _("Rotate"),
+                callback = function()
+                    self.rotated = not self.rotated and true or false
+                    self:update()
+                end,
             },
-            button_table,
-        }
-        button_table_size = button_table:getSize().h
+            {
+                id = "close",
+                text = _("Close"),
+                callback = function()
+                    self:onClose()
+                end,
+            },
+        },
+    }
+    self.button_table = ButtonTable:new{
+        width = self.width - 2*self.button_padding,
+        button_font_face = "cfont",
+        button_font_size = 20,
+        buttons = buttons,
+        zero_sep = true,
+        show_parent = self,
+    }
+    self.button_container = CenterContainer:new{
+        dimen = Geom:new{
+            w = self.width,
+            h = self.button_table:getSize().h,
+        },
+        self.button_table,
+    }
+
+    if self.buttons_visible then
+        self.button_table_size = self.button_table:getSize().h
+    else
+        self.button_table_size = 0
     end
 
     -- height available to our image
-    local img_container_h = self.height - button_table_size
+    self.img_container_h = self.height - self.button_table_size
+
+    -- Iniit the title bar and its components no matter what
+    -- Toggler (white arrow) for caption, on the left of title
+    local ctoggler_text
+    if self.caption_visible then
+        ctoggler_text = "▽ " -- white arrow (nicer than smaller black arrow ▼)
+    else
+        ctoggler_text = "▷ " -- white arrow (nicer than smaller black arrow ►)
+    end
+    self.ctoggler_tw = TextWidget:new{
+        text = ctoggler_text,
+        face = self.title_face,
+    }
```

Comment:
Yeah, it was easier to just initialize everything, and only slot them in the shown container/frame as-needed on update.

But you're entirely right that this means some of those might be orphan at CloseWidget time, which is a problem.
I'll add a manual free pass on those where it matters.

Reply:
> which is a problem.

Not a huge one, just a possible added delay before GC. I think once the widget is closed, they'll be orphaned and GC'ed, and XText should free its stuff.
I think you fixed that recently also for Images/blitbuffer if there are any in there, so it might be fine.
(But of course, I prefer explicite :free() where we can :)